### ISSUE TITLE
What if kprobe unavaliable on my system?

### ISSUE THREAD

**Issue by w296488320** [WeiJiLab/kernel-hook-framework#5]
普通的内核，没开启kprobe ，但是开启了符号 
这个框架底层是inline完全实现的还是通过kprobe 去实现的 ，如果能有更多的技术实现说明就好了。
我希望能找个存inlinehook实现的内核hook框架，这样可以在不开启kprobe 的情况下去使用这个inlinehook

**Comment by liutgnu** [comment]
I guess other people may have the same question, so I rephrase yours:

**_Is it possible to get rid of kprobe in order to use this inline hook framework?_**

The answer is, it may or may not, depending on your kernel version and your familiarity to the kernel. Old kernels have a function called "kallsyms_lookup_name" avaliable, and for the inline hook framework, it mainly use this function to get the addresses of arbitrary kernel functions. Newer kernels no longer export the function, see link [1], so we need to use another way to get the "kallsyms_lookup_name" function re-avaliable. Currently I use kprobes to get it, see src/framework/symbol_resolver.c:init_kallsyms_lookup_func().

So in short, if the kernel you are dealing with, is an old kernel with the function exported, then you can get rid of kprobe with no problem. If not, you can use kprobe, or any other hacking method, depending on your kernel knowledge, to bypass the restriction and regain the "kallsyms_lookup_name" function.

[1]: https://github.com/xcellerator/linux_kernel_hacking/issues/3

**Comment by liutgnu** [comment]
Sorry, I forget to answer your 2nd question:

**_Are there some tech details I can take reference to know its internals?_**

I guess there are tons of documents which describe the mechanism of inline hook better than me. You can take any of these to get the idea of inline hook. However if you like, you can take a look at https://gist.github.com/liutgnu, which contains a simplified arm/aarch64 userspace programs by which I testing inline hook before working on the kernel ones, and that is the very key part about inline hook. You can play around with it before dive into the kernel code.

Besides the key part of inline hook, there are some servicing related functions which will interact with kernel, such as removing kernel code read only permission, managing multiple hooking targets, getting arbitrary kernel symbols' address, kernel/user space interactions etc, nothing special actually.
